include/exadg/operators/structure_calculators.cpp

```diff
@@ -90,10 +90,113 @@ DisplacementJacobianCalculator<dim, Number>::cell_loop(
+{
+  matrix_free              = &matrix_free_in;
+  dof_index_vector         = dof_index_vector_in;
+  dof_index_scalar         = dof_index_scalar_in;
+  quad_index               = quad_index_in;
+  elasticity_operator_base = &elasticity_operator_base_in;
+}
+
+template<int dim, typename Number>
+void
+MaxPrincipalStressCalculator<dim, Number>::compute_projection_rhs(
+  VectorType &       dst_scalar_valued,
+  VectorType const & src_vector_valued) const
+{
+  dst_scalar_valued = 0;
+
+  matrix_free->cell_loop(&This::cell_loop, this, dst_scalar_valued, src_vector_valued);
+}
+
+template<int dim, typename Number>
+void
+MaxPrincipalStressCalculator<dim, Number>::cell_loop(
+  dealii::MatrixFree<dim, Number> const &       matrix_free,
+  VectorType &                                  dst_scalar_valued,
+  VectorType const &                            src_vector_valued,
+  std::pair<unsigned int, unsigned int> const & cell_range) const
+{
+  CellIntegratorVector integrator_vector(matrix_free, dof_index_vector, quad_index, 0);
+  CellIntegratorScalar integrator_scalar(matrix_free, dof_index_scalar, quad_index, 0);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+  {
+    // material_handler->reinit(matrix_free, cell);
+    // std::shared_ptr<Material<dim, Number>> material = material_handler.get_material();
+
+    Structure::Material<dim, Number> const & material =
+      elasticity_operator_base->get_material_in_cell(matrix_free, cell);
+
+    integrator_vector.reinit(cell);
+    // Do not enforce constraints on the `src` vector, as constraints are already applied and the
+    // `dealii::MatrixFree` object might store different constraints.
+    integrator_vector.read_dof_values_plain(src_vector_valued);
+    integrator_vector.evaluate(dealii::EvaluationFlags::gradients);
+
+    integrator_scalar.reinit(cell);
+
+    for(unsigned int q = 0; q < integrator_vector.n_q_points; q++)
+    {
+      tensor const           gradient_displacement = integrator_vector.get_gradient(q);
+      tensor const           F                     = Structure::get_F(gradient_displacement);
+      scalar const           Jacobian              = determinant(F);
+      symmetric_tensor const S =
+        material.second_piola_kirchhoff_stress(gradient_displacement, cell, q);
+      symmetric_tensor const sigma = Structure::compute_push_forward(Jacobian, S, F);
+
+      // Loop over vectorization length to use `dealii::eigenvalues()`.
+      scalar max_eigenvalue;
+      for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); v++)
+      {
+        dealii::SymmetricTensor<2, dim, Number> sigma_v;
+        for(unsigned int i = 0; i < dim; ++i)
+        {
+          for(unsigned int j = 0; j < dim; ++j)
+          {
+            sigma_v[i][j] = sigma[i][j][v];
+          }
+        }
+        std::array<Number, dim> eigenvalues_v = dealii::eigenvalues(sigma_v);
```

Comment:
Yes, this is the right thing to do. Also, doing it with vectorized arrays would require some ternary logic, especially since the code involves some sorting there is some additional action needed.